[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54928 [quant][graphmode][fx][refactor] Quantize by Use of a Tensor instead of Tensor
* **#54924 [quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat**

Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

jerryzh168 committed:
[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat

Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qconcat.cpp:
```diff
@@ -31,6 +31,26 @@ bool is_valid_quantization_scheme(const Tensor& t) {
   return (qtype == kPerTensorAffine) || (qtype == kPerTensorSymmetric);
 }
 
+bool all_inputs_sharing_qparams(TensorList qxs) {
```
is the performance of this neglibigle?

jerryzh168 replied:
metadata heck should be fast, maybe qparam check for per channel quant will be slower, however I think we only have per tensor quant for activations so it should be OK.

Maybe I should error out for the per channel case

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -238,18 +238,20 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
-        scale = float(scale)
-        zero_point = int(zero_point)
+        # scale, zero_point = activation_post_process.calculate_qparams()
+        # scale = float(scale)
+        # zero_point = int(zero_point)
 
-        scale_arg, zero_point_arg = create_qparam_nodes(quantizer, node.name, scale, zero_point)
+        # scale_arg, zero_point_arg = create_qparam_nodes(quantizer, node.name, scale, zero_point)
 
-        kwargs = {**load_arg(quantized=False)(node.kwargs), 'scale': scale_arg, 'zero_point': zero_point_arg}
-        return quantizer.quantized_graph.create_node(
-            'call_function', torch.ops.quantized.cat, load_arg(quantized=[0])(node.args), kwargs)
+        # kwargs = {**load_arg(quantized=False)(node.kwargs), 'scale': scale_arg, 'zero_point': zero_point_arg}
+        # return quantizer.quantized_graph.create_node(
+        #     'call_function', torch.ops.quantized.cat, load_arg(quantized=[0])(node.args), kwargs)
```
nit: remove?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

vkuzo commented:
makes sense.  Does this negatively impact accuracy?  What happens if there is something like this

```
conv1  --- cat
                  /
conv2 --- ---- some_other_nodes
```

and conv1 and conv2 have qparams which are not close to each other?

jerryzh168 commented:
> makes sense. Does this negatively impact accuracy? What happens if there is something like this

probably not, even in current implementation, we add observer for output of cat, which is a concatenation of the output of both convs, so they are observed with the same observer anyways.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]

vkuzo commented on torch/quantization/ns/mappings.py:
```diff
@@ -67,6 +66,7 @@
     torch.avg_pool1d,
     torch._C._nn.avg_pool2d,
     torch._C._nn.avg_pool3d,
+    torch.cat,
```
looks like we also need to update the `TestFXNumericSuiteCoreAPIs.test_op_io_dtype_coverage` test for cat to point to this list of functions, instead of the old one

jerryzh168 committed:
Update on "[quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat"


Summary:
Previously we are producing torch.ops.quantize.cat which takes inputs, dequantize them
and requantize with new qparams. This PR changes that to produce torch.cat directly, torch.cat
will assume all inputs are sharing the same qparam, and it will produce a quantized Tensor with
the same qparam as all inputs (because previous PR makes sure all inputs and output of cat are sharing
the same observer/fakequant instance).

Using torch.cat is expected to be more efficient since it does not introduce extra quant/dequant.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_cat

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27416528](https://our.internmc.facebook.com/intern/diff/D27416528)

[ghstack-poisoned]
